test/distributed_graph_generator_test_utils.h

```diff
@@ -0,0 +1,560 @@
+#pragma once
+
+#include <deque>
+#include <exception>
+#include <functional>
+#include <memory>
+#include <optional>
+#include <vector>
+
+#include <catch2/catch_message.hpp>
+#include <spdlog/fmt/fmt.h>
+
+#include "access_modes.h"
+#include "command_graph.h"
+#include "distributed_graph_generator.h"
+#include "print_graph.h"
+#include "task_manager.h"
+#include "types.h"
+
+#include "test_utils.h"
+
+using namespace celerity;
+using namespace celerity::detail;
+
+class dist_cdag_test_context;
+
+class task_builder {
+	friend class dist_cdag_test_context;
+
+	using action = std::function<void(handler&)>;
+
+	class step {
+	  public:
+		step(dist_cdag_test_context& dctx, std::deque<action> actions) : m_actions(std::move(actions)), m_dctx(dctx) {}
+		virtual ~step() noexcept(false);
+
+		step(const step&) = delete;
+		step(step&&) = delete;
+		step& operator=(const step&) = delete;
+		step& operator=(step&&) = delete;
+
+		task_id submit();
+
+	  private:
+		std::deque<action> m_actions;
+
+	  protected:
+		dist_cdag_test_context& m_dctx;
+
+		template <typename StepT>
+		StepT chain(action a) {
+			static_assert(std::is_base_of_v<step, StepT>);
+			m_actions.push_front(std::move(a));
+			return StepT{m_dctx, std::move(m_actions)};
+		}
+	};
+
+	class buffer_access_step : public step {
```

Comment:
This class hierarchy looks like the remains of some refactoring. All the functions on `buffer_access_step` could (should?) just be members of `step` as far as I can see. This would also address the `TODO misnomer` below.

Reply:
Yes, good point!